README.md
[![CocoaPods Compatible](https://img.shields.io/cocoapods/v/Validated.svg?style=flat-square)](https://cocoapods.org/pods/Validated) [![Platform support](https://img.shields.io/badge/platform-ios%20%7C%20osx%20%7C%20tvos%20%7C%20watchos-lightgrey.svg?style=flat-square)](https://github.com/Ben-G/Validated/blob/master/LICENSE.md)[![Carthage compatible](https://img.shields.io/badge/Carthage-compatible-4BC51D.svg?style=flat)](https://github.com/Carthage/Carthage) [![License MIT](https://img.shields.io/badge/license-MIT-blue.svg?style=flat-square)](https://github.com/Ben-G/Validated/blob/master/LICENSE.md)

#Validated

Validated is a μ-library (~50 Source Lines of Code) that allows you make better use of Swift's type system by providing tools for easily generating new types with built-in guarantees.

Validated allows you to use the type system to verify properties of your values, providing a new level of compile time guarantees.

Using validators you can define new types that add guarantees to existing types:

```swift
// Create a new string type that can never be empty
typealias NonEmptyString = Validated<String, NonEmptyStringValidator>
```

##Example

You might have a function in your code that only knows how to work with a `User` value when the user is logged in. Usually you will implement this requirement in code & add documentation, but you don't have an easy way of expressing this invariant in the type signature:

```swift
/// Please ever only call with a logged-in user!
func performTaskWithUser(user: User) {
    precondition(
    	user.loggedIn, 
    	"It is illegal to call this method with a logged out user!"
    )

	// ...
}
```

Using Validated you can quickly create a new type that describes this requirement in the type system. That makes it impossible to call the function with a logged-out user and it makes the method signature express your invariant (instead of relying on documentation):

```swift
func performTaskWithUser(user: LoggedInUser) {
	// ...
}
```

**So how is this new `LoggedInUser` type created?**

First, you need to implement a validator:

```swift
struct LoggedInValidator: Validator {

    static func validate(value: User) -> Bool {
        return value.loggedIn
    }

}
```
A `Validator` needs to implement the `validate` function that takes the type that this validator can validate (in this case a `User`). The funtion returns a `Bool`. Return `true` if the requirements are fulfilled and `false` if not.

With the `Validator` in place we can create our new type like this:

```swift
typealias LoggedInUser = Validated<User, LoggedInValidator>
```

Note, that it is not required to provide a typealias, but for most cases it is recommended.

**And that's it!**

`LoggedInUser` now has a failable initializer that takes a `User`. If the passed in `User` fulfills the logged-in requirement you will have a `LoggedInUser`, otherwise `nil`. Additionally `LoggedInUser` provides a throwing initializer, in case you prefer to handle failed validations as errors instead of `nil` values.

The underlying value (the full `User` value) is stored in the `.value` property of `LoggedInUser`.

##Beyond the Basics

Validated provides some further features that might be non-obvious.

###Composing Validators with Logical Operators

Validated offers `Validator` types for logical operations that allow you to require multiple validations in different combinations. E.g. you can use the `And` validator to require that two requirements must be met for your type to intializer:

```swift
typealias AllCapsNonEmptyString =
            Validated<String, And<NonEmptyStringValidator, AllCapsLatinStringValidator>>
```
`Or` and `Not` are provided as additional validators. You can take a look at the specs for additional examples.

###Generic Validators

A `Validator` can itself be generic. This is useful if you want to provide verifications for a whole category of types. The example validator `NonEmptyCollectionValidator` can be applied to all validator types by using a generic requirement:

```swift
struct NonEmptyCollectionValidator<T: CollectionType>: Validator {
    static func validate(value: T) -> Bool {
        if !value.isEmpty {
            return true
        } else {
            return false
        }
    }
}
```
However, when using this validator to create a type, you will have to specify the exact type of collection you want to validate:

```swift
typealias NonEmptyListOfStrings = Validated<[String], NonEmptyCollectionValidator<[String]>>
```

# Does this Library Enable Dependent Types?

No, not really. Dependent types would allow us to define types, solely based on values. This library only allows us to validate if a type is of a specified type `<T>` based on its value and a validator. The value itself doesn't change the type information. 

Truely dependent types would create the following results:
```swift
[1, 2, 3] // type = Array<count:3 type:Int>
[1, 2, 3, 4] // type = Array<count:4 type:Int>
```

Using Validated we can only verify if a type falls into one of our statically defined categories:
```swift
ListOf3([1,2,3]) // type = ListOf3<Array<Int>
ListOf3([1,2,3,4]) // type = nil
```

However, these statically provided checks can still add a lot of value to your code; see the examples above.

#Installation

Validated is available via the usual suspects.

## CocoaPods

You can install Validated via CocoaPods by adding it to your `Podfile`:

	use_frameworks!

	source 'https://github.com/CocoaPods/Specs.git'
	platform :ios, '8.0'
	
	pod 'Validated'

And run `pod install`.

## Carthage

You can install Validated via Carthage by adding the following line to your Cartfile:

    github "Ben-G/Validated"


# Get in touch

If you have any questions, you can find me on twitter [@benjaminencz](https://twitter.com/benjaminencz).

Changelog.md
# 3.0.0

*Released: 10/10/2016*

- Migrated to Swift 3.0 - @Ben-G

#2.0.1

*Released: 05/05/2016*

- Lowered iOS Deployment Target to 8.0 - @loganmoseley
- Update usage of `typealias` to `associatedtype` - @AndrewSB

#2.0
*Released: 03/02/2016*

**Breaking API Changes:**

- Removed `Validated2` and `Validated3` in favor of introduced logical wrapper validators - @Ben-G on suggestion of @tomquist
- Failable initializer of `Validated` now requires explicit `value` argument due to introduction of throwing initializer - @Ben-G

**API Additions:**

- Added wrapper validators for logical operators - @tomquist
- Added throwing initializer for `Validated` - @Ben-G upon suggestion of @radex

**Other Changes:**

- Major Refactoring of `Validated` Type - @dehesa
- Addition of OSX, tvOS and watchOS targets - @dehesae

#1.0
*Released: 02/24/2016*

- Initial Release - @Ben-G

Tests/ValidatedTests.swift
import XCTest
import Validated

/**
 The following tests use multiple example validators to ensure that invalid values
 cannot be constructed, while valid ones can be.
*/
class ValidatedTests: XCTestCase {

    // MARK: Basic Validate Tests

    func testValidatesEmptyString() {
        // Define a type for a non-empty string
        typealias NonEmptyString = Validated<String, Not<EmptyStringValidator>>
        // Create an empty string
        let valueNotValidated = NonEmptyString(value: "")?.value
        XCTAssertNil(valueNotValidated)

        // Create a non-empty string
        let valueValidated = NonEmptyString(value: "This is OK.")?.value
        XCTAssertEqual(valueValidated, "This is OK.")
    }

    func testValidatesAnyEmptyCollection() {
        // Define a type for a non-empty collection
        typealias NonEmptyListOfStrings = Validated<[String], Not<EmptyCollectionValidator<[String]>>>
        // Create an empty list of strings
        let valueNotValidated = NonEmptyListOfStrings(value: [])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = NonEmptyListOfStrings(value: ["A", "B"])?.value
        XCTAssertEqual(valueValidated!, ["A", "B"])
    }

    func testSumLarger20() {
        // Define a type for an array of ints that has a sum larger 20
        typealias SumLarger20Array = Validated<[Int], SumLarger20Validator>

        let valueNotValidated = SumLarger20Array(value: [8,8])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = SumLarger20Array(value: [8,8,8])?.value
        XCTAssertEqual(valueValidated!, [8,8,8])
    }

    // MARK: Generic Validator Tests

    func testValidatesCountGreater10() {
        // Define a type for a collection with more than 10 elements
        typealias MoreThan10ElementsIntArray = Validated<[Int], CountGreater10Validator<[Int]>>

        let valueNotValidated = MoreThan10ElementsIntArray(value: [1,2,3,4,5,6])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = MoreThan10ElementsIntArray(value: [1,2,3,4,5,6,7,8,9,10,11])?.value
        XCTAssertEqual(valueValidated!, [1,2,3,4,5,6,7,8,9,10,11])
    }

    // MARK: Custom Type Tests

    func testValidatesLoggedInUser() {
        // Define a type for a logged in user
        typealias LoggedInUser = Validated<User, LoggedInValidator>

        let valueNotValidated = LoggedInUser(
            value: User(username: "User", loggedIn: false)
        )?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = LoggedInUser(
            value: User(username: "User", loggedIn: true)
        )?.value
        XCTAssertEqual(valueValidated?.username, "User")
    }

    // MARK: 2 Validator Test

    func testValidatesEmptyStringAndAllCaps1() {
        // Define a type for a non-empty all caps latin string
        typealias AllCapsNonEmptyString =
            Validated<String, And<Not<EmptyStringValidator>, AllCapsLatinStringValidator>>

        // Create an empty string
        let valueNotValidated = AllCapsNonEmptyString(value: "")?.value
        XCTAssertNil(valueNotValidated)

        // Create a string with lowercase letters
        let valueNotValidated2 = AllCapsNonEmptyString(value: "xYZcF")?.value
        XCTAssertNil(valueNotValidated2)

        // Create an all caps string
        let valueValidated = AllCapsNonEmptyString(value: "XYZCF")?.value
        XCTAssertEqual(valueValidated, "XYZCF")
    }

    func testValidatesCountGreater10AndSumLarger20() {
        typealias MoreThan10ElementsWithSumGreater20 =
            Validated<[Int], And<CountGreater10Validator<[Int]>, SumLarger20Validator>>

        let valueNotValidated = MoreThan10ElementsWithSumGreater20(value: [100,2,400,5,6])?.value
        XCTAssertNil(valueNotValidated)

        let valueNotValidated2 = MoreThan10ElementsWithSumGreater20(
            value: [1,1,1,1,1,1,1,1,1,1,1,1,1,1]
        )?.value
        XCTAssertNil(valueNotValidated2)

        let valueValidated = MoreThan10ElementsWithSumGreater20(
            value: [100,1,1,1,1,1,1,1,1,1,1,1,1,100]
        )?.value
        XCTAssertEqual(valueValidated!, [100,1,1,1,1,1,1,1,1,1,1,1,1,100])
    }

    // MARK: 3 Validator Test

    func testValidatesEmptyStringAndAllCapsContainsYorZ() {
        typealias AllCapsNonEmptyStringWithYorZ =
            Validated<String, And<And<Not<EmptyStringValidator>, AllCapsLatinStringValidator>, ContainsYorZ>>

        // Create a non-complying string
        let valueNotValidated = AllCapsNonEmptyStringWithYorZ(value: "ABCDEF")?.value
        XCTAssertNil(valueNotValidated)

        // Create a compying string
        let valueValidated = AllCapsNonEmptyStringWithYorZ(value: "ABCDEFY")?.value
        XCTAssertEqual(valueValidated, "ABCDEFY")
    }
    
    // MARK: Validator And/Or/Not Tests

    func testValidatesEmptyStringAndAllCaps2() {
        // Define a type for a non-empty all caps latin string
        typealias AllCapsNonEmptyString =
            Validated<String, And<Not<EmptyStringValidator>, AllCapsLatinStringValidator>>

        // Create an empty string
        let valueNotValidated = AllCapsNonEmptyString(value: "")?.value
        XCTAssertNil(valueNotValidated)

        // Create a string with lowercase letters
        let valueNotValidated2 = AllCapsNonEmptyString(value: "xYZcF")?.value
        XCTAssertNil(valueNotValidated2)

        // Create an all caps string
        let valueValidated = AllCapsNonEmptyString(value: "XYZCF")?.value
        XCTAssertEqual(valueValidated, "XYZCF")
    }
    
    func testValidatesEmptyStringOrAllCaps() {
        // Define a type for an empty or all caps latin string
        typealias AllCapsOrEmptyString =
            Validated<String, Or<EmptyStringValidator, AllCapsLatinStringValidator>>
        
        // Create an empty string
        let valueValidated1 = AllCapsOrEmptyString(value: "")?.value
        XCTAssertEqual(valueValidated1, "")
        
        // Create a string with lowercase letters
        let valueNotValidated = AllCapsOrEmptyString(value: "xYZcF")?.value
        XCTAssertNil(valueNotValidated)
        
        // Create an all caps string
        let valueValidated2 = AllCapsOrEmptyString(value: "XYZCF")?.value
        XCTAssertEqual(valueValidated2, "XYZCF")
    }
    
    func testValidatesNonEmpty() {
        // Define a type for a non-empty string
        typealias AllCapsNonEmptyString =
            Validated<String, Not<EmptyStringValidator>>
        
        // Create an empty string
        let valueNotValidated = AllCapsNonEmptyString(value: "")?.value
        XCTAssertNil(valueNotValidated)
        
        // Create an all caps string
        let valueValidated = AllCapsNonEmptyString(value: "abC")?.value
        XCTAssertEqual(valueValidated, "abC")
    }

    // MARK: Error Handling Tests

    func testThrowingInitializer() {
        // Define a type for a non-empty string
        typealias NonEmptyString = Validated<String, Not<EmptyStringValidator>>
        // Create an empty string
        do {
            _ = try NonEmptyString("").value
        } catch let error as ValidatorError {
            XCTAssertTrue(error.validator == Not<EmptyStringValidator>.self)
            XCTAssertTrue(error.wrapperValue as! String == "")
            XCTAssertEqual(
                error.description,
                "Value: '' <String>, failed validation of Validator: Not<EmptyStringValidator>"
            )
        } catch {}

        // Create a non-empty string
        let valueValidated = try! NonEmptyString("This is OK.").value
        XCTAssertEqual(valueValidated, "This is OK.")
    }
}


Tests/helpers/CustomTypeExample.swift
//
//  CustomTypeExample.swift
//  Validated
//
//  Created by Benji Encz on 2/24/16.
//  Copyright © 2016 Benjamin Encz. All rights reserved.
//

import Validated

// Example of a validator that validates a custom type.

struct User {
    let username: String
    let loggedIn: Bool
}

struct LoggedInValidator: Validator {

    static func validate(_ value: User) -> Bool {
        return value.loggedIn
    }

}


Tests/helpers/ExampleValidators.swift
//
//  ExampleValidators.swift
//  Validated
//
//  Created by Benji Encz on 2/24/16.
//  Copyright © 2016 Benjamin Encz. All rights reserved.
//

@testable import Validated

// Example validators that are used throughout the unit tests. These should also be a good starting point for your custom `Validator` types.

struct EmptyStringValidator: Validator {
    static func validate(_ value: String) -> Bool {
        return value.isEmpty
    }
}

struct AllCapsLatinStringValidator: Validator {
    static func validate(_ value: String) -> Bool {
        return value.characters.reduce(true) { accumulator, character in
            return accumulator && ("A"..."Z").contains(character)
        }
    }
}

struct ContainsYorZ: Validator {
    static func validate(_ value: String) -> Bool {
        return value.characters.reduce(false) { accumulator, character in
            return accumulator || ("Y"..."Z").contains(character)
        }
    }
}

struct EmptyCollectionValidator<T: Collection>: Validator {
    static func validate(_ value: T) -> Bool {
        return value.isEmpty
    }
}

struct CountGreater10Validator<T: Collection>: Validator {
    static func validate(_ value: T) -> Bool {
        return value.count > 10
    }
}

struct SumLarger20Validator: Validator {
    static func validate(_ value: [Int]) -> Bool {
        return value.reduce(0, +) > 20
    }
}


Validated/Validated.swift
/// This protocol needs to be implemented in order to add a requirement to
/// a wrapped type.
/// Implementers receive the wrapped type and need to determine if its values
/// fulfill the requirements of the wrapper type.
/// ~~~
/// struct NonEmptyStringValidator: Validator {
///     static func validate(value: String) -> Bool {
///         return !value.isEmpty
///     }
/// }
/// ~~~
public protocol Validator {
    associatedtype WrappedType

    /// Validates if a value of the wrapped type fullfills the requirements of the
    /// wrapper type.
    ///
    /// - parameter validate: An instance of the `WrappedType`  
    /// - returns: A `Bool` indicating success(`true`)/failure(`false`) of the validation
    static func validate(_ value: WrappedType) -> Bool
}


/// Error that is thrown when a validation fails. Proivdes the validator type and 
/// the value that failed validation
public struct ValidatorError: Error, CustomStringConvertible {
    /// The value that failed validation.
    public let wrapperValue: Any
    /// Type of a specific `Validator`. `Any` is used because `Validator` has associated type requirements.
    public let validator: Any.Type

    public var description: String {
        return "Value: '\(wrapperValue)' <\(type(of: (wrapperValue)))>, failed validation of Validator: \(validator.self)"
    }
}

/// Wraps a type together with one validator. Provides a failable initializer
/// that will only return a value of `Validated` if the provided `WrapperType` value
/// fulfills the requirements of the specified `Validator`.
public struct Validated<WrapperType, V: Validator> where V.WrappedType == WrapperType {
    /// The value that passes the defined `Validator`.
    ///
    /// If you are able to access this property; it means the wrappedType passes the validator.
    public let value: WrapperType

    /// Throwing initializer that will *not* throw an error if the provided value fulfills the requirements 
    /// specified by the `Validator`.
    public init(_ value: WrapperType) throws {
        guard V.validate(value) else {
            throw ValidatorError(
                wrapperValue: value,
                validator: V.self)
        }

        self.value = value
    }

    /// Failible initializer that will only succeed if the provided value fulfills the requirements specified by the `Validator`.
    public init?(value: WrapperType) {
        try? self.init(value)
    }
}

/// Validator wrapper which is valid when `V1` and `V2` validated to `true`.
public struct And<
    V1: Validator,
    V2: Validator>: Validator where
        V1.WrappedType == V2.WrappedType {
    public static func validate(_ value: V1.WrappedType) -> Bool {
        return V1.validate(value) && V2.validate(value)
    }
}

/// Validator wrapper which is valid when either `V1` or `V2` validated to `true`.
public struct Or<
    V1: Validator,
    V2: Validator>: Validator where
        V1.WrappedType == V2.WrappedType {
    public static func validate(_ value: V1.WrappedType) -> Bool {
        return V1.validate(value) || V2.validate(value)
    }
}

/// Validator wrapper which is valid when `V1` validated to `false`.
public struct Not<V1: Validator>: Validator {
    public static func validate(_ value: V1.WrappedType) -> Bool {
        return !V1.validate(value)
    }
}


ValidatedTests/TestSupport/ExampleValidators.swift
//
//  ExampleValidators.swift
//  Validated
//
//  Created by Benji Encz on 2/24/16.
//  Copyright © 2016 Benjamin Encz. All rights reserved.
//

@testable import Validated

/**
 Example validators that are used throughout the unit tests.
 These should also be a good starting point for your custom `Validator` types.
 */

struct EmptyStringValidator: Validator {
    static func validate(value: String) -> Bool {
        return value.isEmpty
    }
}

struct AllCapsLatinStringValidator: Validator {
    static func validate(value: String) -> Bool {
        return value.characters.reduce(true) { accumulator, character in
            return accumulator && ("A"..."Z").contains(character)
        }
    }
}

struct ContainsYorZ: Validator {
    static func validate(value: String) -> Bool {
        return value.characters.reduce(false) { accumulator, character in
            return accumulator || ("Y"..."Z").contains(character)
        }
    }
}

struct EmptyCollectionValidator<T: CollectionType>: Validator {
    static func validate(value: T) -> Bool {
        return value.isEmpty
    }
}

struct CountGreater10Validator<T: CollectionType>: Validator {
    static func validate(value: T) -> Bool {
        return value.count > 10
    }
}

struct SumLarger20Validator: Validator {
    static func validate(value: [Int]) -> Bool {
        return value.reduce(0, combine: +) > 20
    }
}


ValidatedTests/ValidatedTests.swift
//
//  ValidatedTests.swift
//  ValidatedTests
//
//  Created by Benji Encz on 2/24/16.
//  Copyright © 2016 Benjamin Encz. All rights reserved.
//

import XCTest
import Validated

/**
 The following tests use multiple example validators to ensure that invalid values
 cannot be constructed, while valid ones can be.
*/
class ValidatedTests: XCTestCase {

    // MARK: Basic Validate Tests

    func testValidatesEmptyString() {
        // Define a type for a non-empty string
        typealias NonEmptyString = Validated<String, NonEmptyStringValidator>
        // Create an empty string
        let valueNotValidated = NonEmptyString("")?.value
        XCTAssertNil(valueNotValidated)

        // Create a non-empty string
        let valueValidated = NonEmptyString("This is OK.")?.value
        XCTAssertEqual(valueValidated, "This is OK.")
    }

    func testValidatesAnyEmptyCollection() {
        // Define a type for a non-empty collection
        typealias NonEmptyListOfStrings = Validated<[String], NonEmptyCollectionValidator<[String]>>
        // Create an empty list of strings
        let valueNotValidated = NonEmptyListOfStrings([])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = NonEmptyListOfStrings(["A", "B"])?.value
        XCTAssertEqual(valueValidated!, ["A", "B"])
    }

    func testSumLarger20() {
        // Define a type for an array of ints that has a sum larger 20
        typealias SumLarger20Array = Validated<[Int], SumLarger20Validator>

        let valueNotValidated = SumLarger20Array([8,8])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = SumLarger20Array([8,8,8])?.value
        XCTAssertEqual(valueValidated!, [8,8,8])
    }

    // MARK: Generic Validator Tests

    func testValidatesCountGreater10() {
        // Define a type for a collection with more than 10 elements
        typealias MoreThan10ElementsIntArray = Validated<[Int], CountGreater10Validator<[Int]>>

        let valueNotValidated = MoreThan10ElementsIntArray([1,2,3,4,5,6])?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = MoreThan10ElementsIntArray([1,2,3,4,5,6,7,8,9,10,11])?.value
        XCTAssertEqual(valueValidated!, [1,2,3,4,5,6,7,8,9,10,11])
    }

    // MARK: Custom Type Tests

    func testValidatesLoggedInUser() {
        // Define a type for a logged in user
        typealias LoggedInUser = Validated<User, LoggedInValidator>

        let valueNotValidated = LoggedInUser(
            User(username: "User", loggedIn: false)
        )?.value
        XCTAssertNil(valueNotValidated)

        let valueValidated = LoggedInUser(
            User(username: "User", loggedIn: true)
        )?.value
        XCTAssertEqual(valueValidated?.username, "User")
    }

    // MARK: Validate2 Tests

    func testValidatesEmptyStringAndAllCaps() {
        // Define a type for a non-empty all caps latin string
        typealias AllCapsNonEmptyString =
            Validated2<String, NonEmptyStringValidator, AllCapsLatinStringValidator>

        // Create an empty string
        let valueNotValidated = AllCapsNonEmptyString("")?.value
        XCTAssertNil(valueNotValidated)

        // Create a string with lowercase letters
        let valueNotValidated2 = AllCapsNonEmptyString("xYZcF")?.value
        XCTAssertNil(valueNotValidated2)

        // Create an all caps string
        let valueValidated = AllCapsNonEmptyString("XYZCF")?.value
        XCTAssertEqual(valueValidated, "XYZCF")
    }

    func testValidatesCountGreater10AndSumLarger20() {
        typealias MoreThan10ElementsWithSumGreater20 =
            Validated2<[Int], CountGreater10Validator<[Int]>, SumLarger20Validator>

        let valueNotValidated = MoreThan10ElementsWithSumGreater20([100,2,400,5,6])?.value
        XCTAssertNil(valueNotValidated)

        let valueNotValidated2 = MoreThan10ElementsWithSumGreater20(
            [1,1,1,1,1,1,1,1,1,1,1,1,1,1]
        )?.value
        XCTAssertNil(valueNotValidated2)

        let valueValidated = MoreThan10ElementsWithSumGreater20(
            [100,1,1,1,1,1,1,1,1,1,1,1,1,100]
        )?.value
        XCTAssertEqual(valueValidated!, [100,1,1,1,1,1,1,1,1,1,1,1,1,100])
    }

    // MARK: Validate3 Tests

    func testValidatesEmptyStringAndAllCapsContainsYorZ() {
        typealias AllCapsNonEmptyStringWithYorZ =
            Validated3<String, NonEmptyStringValidator, AllCapsLatinStringValidator, ContainsYorZ>

        // Create a non-complying string
        let valueNotValidated = AllCapsNonEmptyStringWithYorZ("ABCDEF")?.value
        XCTAssertNil(valueNotValidated)

        // Create a compying string
        let valueValidated = AllCapsNonEmptyStringWithYorZ("ABCDEFY")?.value
        XCTAssertEqual(valueValidated, "ABCDEFY")
    }

}
